### ISSUE TITLE
Pay to pubkey hash

### ISSUE THREAD

**Issue by Insoleet** [duniter/duniter#1002]
I think it could be interesting to send money to pubkey hash instead of directly the pubkey. This is what bitcoin calls "addresses".

This would be a first step in anonymization : we could send money to a second wallet, but this second wallet would be hidden from the public until some money is spent from it.

**Comment by c-geek** [comment]
Yes. It seems quite easy to do also: we would add a lock function:

    P2PKH(ADDRESS)

Then in a transaction spending this source, the unlock would be:

    P2PKH(0)

Where `0` references an issuer of the transaction. So we have in the end:

    P2PKH(0) => P2PKH(PUBKEY)

The condition `P2PKH(0)` would be fulfilled if:

* `HASH(PUBKEY) == ADDRESS`
* `SIG(PUBKEY) == true`

----

Example:

* the HASH function is SHA256
* PUBKEY = 97s9n5sWvhfcZX7is6cXi3nYUxwBgbbnDJVanomuhCMC

We have the lock condition:

    P2PKH(9d7a356d8ebf4301244c7f6fb86ad0f327ef94ba2da17c9bcfc11ac346449a0f)

-----

Also, the HASH function could be a more complex function like Bitcoin is doing:

https://en.bitcoin.it/wiki/Technical_background_of_version_1_Bitcoin_addresses

**Comment by Insoleet** [comment]
The algorithm used by Bitcoin is interesting, especially the RIPEMD-160 hashing which shortens the address.

But I wonder why they do that much hashing in a row. This increase, I think, the number of possible collisions. Do you know why they generate it like this ?

**Comment by Insoleet** [comment]
I found the answer. It is quite interesting and we can probably reproduce their idea : 

http://procbits.com/2013/08/27/generating-a-bitcoin-address-with-javascript

> Version = 1 byte of 0 (zero); on the test network, this is 1 byte of 111
Key hash = Version concatenated with RIPEMD-160(SHA-256(public key))
Checksum = 1st 4 bytes of SHA-256(SHA-256(Key hash))
Bitcoin Address = Base58Encode(Key hash concatenated with Checksum)

**Comment by c-geek** [comment]
I'm adding this issue to 1.4.0 milestone. Feel free to continue your reflexion, this version has no due date yet.

**Comment by c-geek** [comment]
Also see https://forum.duniter.org/t/protocole-paiement-vers-une-adresse/2730 for more infos.

**Comment by librelois** [comment]
I propose to implement P2PKH for version 2.0 at the same time as P2SH
link to #1165
